req-manager.js

```diff
@@ -0,0 +1,393 @@
+    for (const [feedId, hops] of this._requested) {
+      this._requestables.set(feedId, hops)
+      this._requested.delete(feedId)
+    }
+    for (const [feedId, hops] of this._requestedPartially) {
+      this._requestables.set(feedId, hops)
+      this._requestedPartially.delete(feedId)
+    }
+    if (this._liveDrainer) this._liveDrainer.abort()
+    this._liveDrainer = null
+    if (this._liveDrainer2) this._liveDrainer2.abort()
+    this._liveDrainer2 = null
+    this._flush()
+  }
+
+  _setupTemplates(optsPartialReplication) {
+    if (!optsPartialReplication) return null
+    if (Object.values(optsPartialReplication).every((t) => !t)) return null
+    const hopsArr = Object.keys(optsPartialReplication)
+      .map(Number)
+      .filter((x) => x >= 0)
+    const templates = new Map()
+    for (const hops of hopsArr) {
+      if (optsPartialReplication[hops]) {
+        templates.set(hops, new Template(optsPartialReplication[hops]))
+      } else {
+        templates.set(hops, null)
+      }
+    }
+    return templates
+  }
+
+  _someTemplate(fn) {
+    if (!this._templates) return false
+    return [...this._templates.values()].filter((t) => !!t).some(fn)
+  }
+
+  _findTemplateForHops(hops) {
+    if (!this._templates) return null
+    const eligibleHopsArr = [...this._templates.keys()].filter((h) => h >= hops)
+    const pickedHops = Math.min(...eligibleHopsArr)
+    return this._templates.get(pickedHops)
+  }
+
+  _setupCloseHook() {
+    this._hasCloseHook = true
+    const that = this
+    this._ssb.close.hook(function (fn, args) {
+      if (that._liveDrainer) that._liveDrainer.abort()
+      if (that._liveDrainer2) that._liveDrainer2.abort()
+      fn.apply(this, args)
+    })
+  }
+
+  _getCurrentHops(feedId) {
+    let h
+    h = this._requestables.get(feedId)
+    if (typeof h === 'number') return h
+    h = this._requested.get(feedId)
+    if (typeof h === 'number') return h
+    h = this._requestedPartially.get(feedId)
+    if (typeof h === 'number') return h
+    return null
+  }
+
+  _scanBendyButtFeeds() {
+    if (this._liveDrainer) return
+    if (!this._hasCloseHook) this._setupCloseHook()
+
+    pull(
+      this._ssb.metafeeds.branchStream({ old: true, live: true }),
+      (this._liveDrainer = pull.drain((branch) => {
+        const metaFeedId = branch[0][0]
+        const mainFeedId = this._metafeedFinder.getInverse(metaFeedId)
+        this._handleBranch(branch, mainFeedId)
+      }))
+    )
+
+    // Automatically switch to partial replication if (while replicating fully)
+    // we bump into a metafeed/announce msg
+    pull(
+      this._metafeedFinder.liveStream(),
+      (this._liveDrainer2 = pull.drain(([mainFeedId, metaFeedId]) => {
+        if (
+          this._requested.has(mainFeedId) &&
+          !this._requestedPartially.has(metaFeedId)
+        ) {
+          debug(
+            'switch from full replication to partial replication for %s',
```

Comment:
I'm thinking about this one. On one hand it is great that this is handled as it is most correct, on the other hand this should only happen if getSubset doesn't give you anything and you replicate fully and the announce messages gets indexed before you finish the main feed in full?

Reply:
The other case where this happens is when all your currently-connected peers don't have the `getSubset` RPC

Reply:
The main feed will be replicated in "full" but at least from one point onwards, it will default to partial replication

Reply:
ah yes, I was mostly thinking about some initial replication, but you are of course correct in that this is a bit more general. Great!